### ISSUE TITLE
Errors While Decoding Response Text Using mitmdump

### ISSUE THREAD

**Issue by omarmhaimdat** [mitmproxy/mitmproxy#4415]
#### Problem Description
When I visit websites with Arabic characters using a small addon script with mitmdump, and extract the response text I get the following error:

`Traceback (most recent call last):
  File "main.py", line 36, in response
    response_text = flow.response.text
  File "c:\users\evead-61\appdata\local\programs\python\python38\lib\site-packages\mitmproxy\net\http\message.py", line 232, in get_text
    return cast(str, encoding.decode(content, enc))
  File "c:\users\evead-61\appdata\local\programs\python\python38\lib\site-packages\mitmproxy\net\http\encoding.py", line 76, in decode
    raise ValueError("{} when decoding {} with {}: {}".format(
ValueError: UnicodeDecodeError when decoding b'GIF89a\x with 'UTF-8': UnicodeDecodeError('utf-8', b'GIF89a\x01\x00\x01\x00\xf0\x00\x00\x00\x00\x00\x00\x00\x00!\xf9\x04\x01\x00\x00\x00\x00,\x00\x00\x00\x00\x01\x00\x01\x00\x00\x02\x02D\x01\x00;`

#### Steps to reproduce the behavior:
1. Write a small addon that assigns the HTTPResponse flow response text from the "response()" method
2. Assign _flow.response.text_ to a variable
3. Run using `mitmdump -s main.py --anticomp` (assuming your file is called _main.py_)
4. You can try it on this website [chouftv.ma](https://chouftv.ma/press/341484.html)

#### System Information
Paste the output of "mitmproxy --version" here.
Mitmproxy: 6.0.2
Python:    3.8.7
OpenSSL:   OpenSSL 1.1.1i  8 Dec 2020
Platform:  Windows-10-10.0.17763-SP0

**Comment by Prinzhorn** [comment]
> When I visit websites with Arabic characters using a small addon script with mitmdump

Looking at the error you seem to be accessing the `text` property of an _image_ response. ` b'GIF89a\x` are the magic bytes of a gif image. But it shouldn't crash, I need to look into why our guess_encoding thinks this is UTF-8 (should fall back to `latin-1` for binary bodies).

In the meantime, could you test with the latest master? Snapshots are here https://mitmproxy.org/downloads/#branches/master/

**Comment by Prinzhorn** [comment]
I'm able to reproduce this on github.com with their tracking pixel, so it's definitely unrelated to Arabic

**Comment by Prinzhorn** [comment]
ouch oof owie

```
content-type: image/gif; charset=utf-8
```

this comes from the https://collector.githubapp.com/github/page_view tracking pixel

![Selection_795](https://user-images.githubusercontent.com/679144/105995384-316e8c00-60a9-11eb-9276-7bfd3af8f2a2.png)

So I guess we need to be more intelligent when doing `guess_encoding`? @mhils 

I was using 

```py
def response(flow):
    print(flow.request.url)
    print(len(flow.response.text))
```

and visiting GitHub will cause (on master)

```
Addon error: Traceback (most recent call last):
  File "/home/alex/Projects/super-top-secret/src/forks/mitmproxy/mitmproxy/net/http/encoding.py", line 67, in decode
    decoded = custom_decode[encoding](encoded)
KeyError: 'utf-8'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/alex/Projects/super-top-secret/src/forks/mitmproxy/mitmproxy/net/http/encoding.py", line 69, in decode
    decoded = codecs.decode(encoded, encoding, errors)  # type: ignore
  File "/usr/lib/python3.8/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 10: invalid start byte

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/alex/Projects/super-top-secret/src/issues/mitmproxy-4415/main.py", line 3, in response
    print(len(flow.response.text))
  File "/home/alex/Projects/super-top-secret/src/forks/mitmproxy/mitmproxy/net/http/message.py", line 244, in get_text
    return cast(str, encoding.decode(content, enc))
  File "/home/alex/Projects/super-top-secret/src/forks/mitmproxy/mitmproxy/net/http/encoding.py", line 76, in decode
    raise ValueError("{} when decoding {} with {}: {}".format(
ValueError: UnicodeDecodeError when decoding b'GIF89a\x with 'utf-8': UnicodeDecodeError('utf-8', b'GIF89a\x01\x00\x01\x00\x80\xff\x00\xff\xff\xff\x00\x00\x00,\x00\x00\x00\x00\x01\x00\x01\x00\x00\x02\x02D\x01\x00;', 10, 11, 'invalid start byte')
```

**Comment by Prinzhorn** [comment]
I did _some_ research last year since my mitmproxy fork maintains a public version of `guess_encoding`. Maybe there is a library we can use? Or some sort of standard? E.g. this is what I found https://dev.w3.org/html5/cts/html5-type-sniffing.html

**Comment by omarmhaimdat** [comment]
Maybe the encoding should be correlated with the content-type

**Comment by mhils** [comment]
Yes, this is a tricky topic - I recall some heated discussions with @cortesi years ago. 😄 

The `.text` property:

- If `text` were a regular function (like this: `flow.response.text()`), I think that there's a strong argument that the current behavior is correct. mitmproxy is doing everything correctly, we raise because of a server error.
- However, `text` isn't a function but a property instead, and programmers have a very different expecation of how that should behave. Long story short, it's unexpected that a property raises on access.

What to do instead? 

- Do a better job at guessing. TL;DR, this may fix some occasions, but doesn't solve the problem. What's the proper encoding of the tracking pixel above? "binary" is not a valid encoding. 😄 
- We can ultimately handle failure in three different ways:
  1. Raise an error (as we do now)
  2. Use a "replace" error handler (this is for example what requests does)
  3. Use a "surrogateescape" error handler

Let me expand on these options a bit: 

Raising has some usability issues, as evidenced here, but at least it rubs them in your face. One mitigation would be to document this better. 

Using a replace error handler has the problem that an innocuous-looking
`
flow.response.text = flow.response.text.replace("foo","bar")
`
would corrupt the tracking pixel mentioned above - a "replace" error handler is not [bijective](https://en.wikipedia.org/wiki/Bijection). Again, this is now how you would expect properties to work.

Finally, we could use the "surrogateescape" escape handler, which basically implements [UTF-8b](https://vstinner.github.io/pep-383.html). While this generally sounds like the most sane approach to me, it comes with the clear caveat that your program may blow up later at a completely different point and it's hard to pinpoint the root cause.

----


To sum it up, none of the solutions are really ideal. I feel that it'd be reasonable to switch to surrogate-escaping, but that solution definitely has its own problems as well.

**Comment by Prinzhorn** [comment]
>  Do a better job at guessing. TL;DR, this may fix some occasions, but doesn't solve the problem. What's the proper encoding of the tracking pixel above? "binary" is not a valid encoding.

I personally prefer this option since the problem cannot really be solved. Being able to replace stuff inside binary bodies is neat, e.g. search & replace meta data in images or pdfs. And I guess `latin-1` gets that job done and keeping it for backwards compat is nice. I would try to expand our heuristics and add new special cases as we find them. I assume that's basically what browser vendors do but by now they've seen 99.9999% of weird shit.

Now the fun begins. We can fall back to latin-1 for `image/*` but not for `image/svg+xml`. Same for `audio/*`, `video/*` and `application/octet-stream`.

If we can agree that this is a valid solution I'll grab a list of common mime types and improve the guessing we currently have.
